Benchmarks/GaussianElimination.cs

```diff
@@ -0,0 +1,278 @@
+                        double maxj = Math.Abs(Ab[i][j]);
+                        if (maxj > max)
+                        {
+                            pi = i;
+                            max = maxj;
+                        }
+                    }
+
+                    // Swap pivot row with the current row.
+                    if (pi != j)
+                    {
+                        var tmp = Ab[pi];
+                        Ab[pi] = Ab[j];
+                        Ab[j] = tmp;
+                    }
+
+                    // Eliminate the rows after the pivot.
+                    double p = Ab[j][j];
+                    for (int i = j + 1; i < M; ++i)
+                    {
+                        double s = Ab[i][j] / p;
+                        if (s != 0.0d)
+                        {
+                            for (int jj = j + 1; jj <= N; ++jj)
+                            {
+                                Ab[i][jj] -= Ab[j][jj] * s;
+                            }
+                        }
+                    }
+                }
+            }
+        }
+
+        [Benchmark]
+        public void ArrayOfArraysVectorized()
+        {
+            for (int iteration = 0; iteration < Size; iteration++)
+            {
+                var Ab = _data[iteration].JaggedArray;
+
+                const double tiny = 0.00001;
+
+                // Solve for dx.
+                // For each variable in the system...
+                for (int j = 0; j + 1 < N; ++j)
+                {
+                    int pi = j;
+                    double max = Math.Abs(Ab[j][j]);
+
+                    // Find a pivot row for this variable.
+                    for (int i = j + 1; i < M; ++i)
+                    {
+                        // if(|JxF[i][j]| > max) { pi = i, max = |JxF[i][j]| }
+                        double maxj = Math.Abs(Ab[i][j]);
+                        if (maxj > max)
+                        {
+                            pi = i;
+                            max = maxj;
+                        }
+                    }
+
+                    // Swap pivot row with the current row.
+                    if (pi != j)
+                    {
+                        var tmp = Ab[pi];
+                        Ab[pi] = Ab[j];
+                        Ab[j] = tmp;
+                    }
+
+                    var vectorLength = System.Numerics.Vector<double>.Count;
+                    // Eliminate the rows after the pivot.
+                    double p = Ab[j][j];
+                    for (int i = j + 1; i < M; ++i)
+                    {
+                        double s = Ab[i][j] / p;
+                        if (Math.Abs(s) >= tiny)
+                        {
+                            int jj;
+                            for (jj = j + 1; jj <= (N - vectorLength); jj += vectorLength)
+                            {
+                                var source = new System.Numerics.Vector<double>(Ab[j], jj);
```

Comment:
I wonder if in order to get the best performance, we need to store the matrix itself as vectors, rather than construct them on the fly?

Reply:
That functionality uses some compiler intrinsics under the hood, so we are not really constructing them here ;)